### ISSUE TITLE
rop.call fails when sigreturn automatically used with strings

### ISSUE THREAD

**Issue by adzy2k6** [Gallopsled/pwntools#1482]
As per the title. If you try to call rop.call('execve', ['/bin/sh']) for example, the rop module will often try to generate the code using srop. The issue is that the string replaces the value of the register in the frame with a reference to the string, and when the rop stack is generated, it pushes the string in multiple words to the stack, rather than setting the register to point to the string.

The issue is these lines in the rop.py file:

```python
for register, value in zip(frame.arguments, arguments):
    frame[register] = value
````
I do intend to have a look at this myself, but I'm limited for time at the moment, and the fix seems simple enough that someone else may want to do it.

**Comment by zachriggle** [comment]
Can you provide more details, an example would help, as well as the expected output.

**Comment by adzy2k6** [comment]
The function I am attempting to call is the execve syscall. The example output is shown below. The offending line appears to be `b'/bin/sh' ebx = b'/bin/sh'`, which it seems to think can be written to ebx directly. What actually happens is the string takes two words on the stack, and gets written to ebx and edx, shuffling the rest of the values down by one word.

```
[*] Using sigreturn for 'SYS_execve'
0xffb3fedc:        0x805c34b pop eax; ret
0xffb3fee0:             0x77 [arg0] eax = SYS_sigreturn
0xffb3fee4:        0x8049a21 int 0x80
0xffb3fee8:              0x0 gs
0xffb3feec:              0x0 fs
0xffb3fef0:              0x0 es
0xffb3fef4:              0x0 ds
0xffb3fef8:              0x0 edi
0xffb3fefc:              0x0 esi
0xffb3ff00:              0x0 ebp
0xffb3ff04:       0xffb3ff38 esp (+0x34)
0xffb3ff08:       b'/bin/sh' ebx = b'/bin/sh'
0xffb3ff0f:              0x0 edx
0xffb3ff13:              0x0 ecx
0xffb3ff17:              0xb eax = SYS_execve
0xffb3ff1b:              0x0 trapno
0xffb3ff1f:              0x0 err
0xffb3ff23:        0x8049a21 int 0x80
0xffb3ff27:             0x23 cs
0xffb3ff2b:              0x0 eflags
0xffb3ff2f:              0x0 esp_at_signal
0xffb3ff33:             0x2b ss
0xffb3ff37:              0x0 fpstate
```

I would expect the actual result to be something closer to this:
```
0xffb3ff04:       0xffb3ff38 esp (+0x34)
0xffb3ff08:       0xffb3ffff  ebx = address of b'/bin/sh'
0xffb3ff0f:              0x0 edx
     ...
0xffb3fff: b'/bin/sh'
```

I validated this by reading the data in the chain() output. It treats the string as binary data to write directly to the register, rather than as a pointer.

**Comment by adzy2k6** [comment]
I fixed this, at least as far as my problem triggered it. The pull request is https://github.com/Gallopsled/pwntools/pull/1484

**Comment by Arusekk** [comment]
Closed in #1484
